### ISSUE TITLE
[Proposal] loadSeedersFrom method in Package service providers

### ISSUE THREAD

**Issue by michaeljberry** [laravel/internals#844]
In an effort to help packages be more flexible and independent, I'd like to propose a `$this->loadSeedersFrom` method that can be included in a package's service provider. 

I imagine the functionality would be similar to `$this->loadMigrationsFrom` where the directory of the package's seeders is passed in.  ex: `$this->loadSeedersFrom(__DIR__ . '/Seeders');`

Then Laravel adds the classes within that directory to the `Seeder->paths` array much like migrations are added. 

Then upon running `migrate` with the `--seed` option or directly running `db:seed` the classes stored in the `paths` array will be `run()` and all seeds will be put into the database.

**Comment by sisve** [comment]
Currently the seeders that are executed by db:seed are decided by their presence in the [DatabaseSeeder](https://github.com/laravel/laravel/blob/master/database/seeds/DatabaseSeeder.php) class. How would I make sure that no unwanted seeders are executed with the described behavior?

I've written my seeders to update-or-create data, they have the expected state for many things (like zip codes), and I can execute the seeders over and over to reset the content to the expected state. Having third party seeders automatically run in my environment, without taking into account foreign keys or existing data, sounds like a recipe for disaster.

**Comment by michaeljberry** [comment]
Good point and I understand that this would likely be a bigger kind of change from at least a workflow standpoint.

On the flip side, though, how is this dissimilar to third party migrations being run in an environment? If I, as the developer, have the `$this->loadMigrationsFrom` in my Service Provider, that would make me solely responsible for any and all migrations that happen. So, if a migration in a third party package, rather than playing nicely, bullies my database and edits or removes columns or tables, I'm in a similar situation.

I understand that the actions are similar, though different. One action (migrations), creates/edits/deletes the environment for data to be stored while the other action actually loads the environment with data. Which action can cause more of a disaster? I've seen the [xkcd with little Bobby Tables]([https://xkcd.com/327/]) and I understand the risks of SQL injection, but havoc could also be had with nefarious migrations being run also. From what I've read and what I understand there's nothing that prevents the `Schema::dropIfExists('users');` function from being called from the `up` method in a migration...

I just suggested the same ease of creating data as we have in creating the environment to house the data - from a third party package. Ultimately, it's still my responsibility to know what seeders/migrations will be run on my project. That never goes away.

I appreciate this discussion and thank you for giving me an opportunity to look at this suggestion from a wider point of view!

**Comment by sisve** [comment]
The largest difference between seeders and migrations is that migrations are only run once, and they have a expected way of working. The up() should create what is needed, the down() should revert it, and Laravel keeps track of which migrations has been executed.

There is no such rules for seeders. Someone seeding a list of all countries can either,

1. insert all rows,
2. delete all rows, insert all new rows,
3. update all existing rows and insert only new data

The first alternative will cause duplicate data, in best case, or failing unique constraints. The second alternative will fail if there are any foreign keys involved since it attempts to delete rows other table references.

I believe that the third alternative is the correct one. It means you can run your seeders over and over again, and always get the seeded tables back to a known state, but it also means that every non-seeded table that references the seeded tables still work too. However, the documentation says that seeders is for "seeding your database with test data", but does not clarify further on how seeders are expected to work if you execute them twice.

I think the described functionality have to be something I can opt-out from. The most simple implementation would be another line of code in the DatabaseSeeder, something like a `Seeder::runAllPackageMigrations()` or perhaps a `Seeder::runMigrationsFrom('sisve/awesome-package');`

At least keep a way or flag somewhere that allows me to opt-out of this new functionality.

**Comment by sagunkho** [comment]
+1 to this idea, it would be really cool for more things to be loaded directly from the package instead of being copied. This is one of the best things about laravel.

**Comment by donnysim** [comment]
In one of my projects I had it prompt a list of seeders where you could select which you want to run, or all of them. With long lists it could be harder to manage, but it did the job. Maybe make it similar? `php artisan db:seed --interactive`.
